src/coreclr/vm/comdelegate.cpp

```diff
@@ -801,49 +898,68 @@ static Stub* SetupShuffleThunk(MethodTable* pDelMT, MethodDesc* pTargetMeth)
     }
     CONTRACTL_END;
 
-    GCX_PREEMP();
-
+    bool isInstRetBuff = (!pTargetMeth->IsStatic() && pTargetMeth->HasRetBuffArg() && IsRetBuffPassedAsFirstArg());
     DelegateEEClass * pClass = (DelegateEEClass *)pDelMT->GetClass();
 
+    // Look for a thunk cached on the delegate class first. Note we need a different thunk for instance methods with a
+    // hidden return buffer argument because the extra argument switches place with the target when coming from the caller.
+    Stub* pShuffleThunk = isInstRetBuff ? pClass->m_pInstRetBuffCallStub : pClass->m_pStaticCallStub;
+    if (pShuffleThunk)
+        return pShuffleThunk->GetEntryPoint();
+
+    GCX_PREEMP();
+
     MethodDesc *pMD = pClass->GetInvokeMethod();
 
+    // We haven't already setup a shuffle thunk, go do it now (which will cache the result automatically).
     StackSArray<ShuffleEntry> rShuffleEntryArray;
-    GenerateShuffleArray(pMD, pTargetMeth, &rShuffleEntryArray);
+    bool isSimple = GenerateShuffleArray(pMD, pTargetMeth, &rShuffleEntryArray);
+    if (isSimple)
```

Comment:
```suggestion
    if (GenerateShuffleArray(pMD, pTargetMeth, &rShuffleEntryArray))
```
We can just check `HasExternalEntryPoint` so that the code is identical to the one in destructor.

Reply:
ok